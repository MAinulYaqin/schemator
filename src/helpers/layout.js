/**
 * Get relation line path points (for SVG d attribute)
 *
 * @param {object} fromTablePosition Position of table which has foreign key
 * @param {number} fromTablePosition.x Position X
 * @param {number} fromTablePosition.y Position Y
 * @param {object} toTablePosition Position of destination table
 * @param {number} toTablePosition.x Position X
 * @param {number} toTablePosition.y Position Y
 * @param {number} fieldIndex Index of foreign key field relative from it's table
 * @returns {string} Path points
 */
export const getPathPoints = (
  fromTablePosition,
  toTablePosition,
  fieldIndex
) => {
  const { x: fPosX, y: fPosY } = fromTablePosition;
  const { x: tPosX, y: tPosY } = toTablePosition;
  const headerHeight = 36;
  const inputHeight = 38;
  const tableWidth = 240;
  const curvePoint = 64;
  const halfHeaderHeight = headerHeight / 2;
  const inputY = headerHeight + fieldIndex * inputHeight + inputHeight / 2;
  const isFromTableInRight = fPosX > tPosX;
  let points = "";

  if (isFromTableInRight) {
    points = `
        M${fPosX + halfHeaderHeight} ${fPosY + inputY} 
        C${fPosX - curvePoint} ${fPosY + inputY} 
          ${tPosX + tableWidth + curvePoint} ${tPosY + halfHeaderHeight} 
          ${tPosX + tableWidth} ${tPosY + halfHeaderHeight}
      `;
  } else {
    points = `
        M${fPosX + tableWidth - halfHeaderHeight} ${fPosY + inputY} 
        C${fPosX + tableWidth + curvePoint} ${fPosY + inputY} 
          ${tPosX - curvePoint} ${tPosY + halfHeaderHeight} 
          ${tPosX + halfHeaderHeight} ${tPosY + halfHeaderHeight}
      `;
  }

  return points;
};
