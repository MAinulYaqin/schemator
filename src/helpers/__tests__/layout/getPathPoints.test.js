import { getPathPoints } from "../../layout";

describe("getPathPoints()", () => {
  it("should returns path points (foreign key in right table)", () => {
    const fromPosition = { x: 320, y: 32 };
    const toPosition = { x: 32, y: 32 };
    const fieldIndex = 0;
    const result = getPathPoints(fromPosition, toPosition, fieldIndex);

    expect(result).toMatchSnapshot();
  });

  it("should returns path points (foreign key in left table)", () => {
    const fromPosition = { x: 32, y: 32 };
    const toPosition = { x: 320, y: 32 };
    const fieldIndex = 0;
    const result = getPathPoints(fromPosition, toPosition, fieldIndex);

    expect(result).toMatchSnapshot();
  });
});
