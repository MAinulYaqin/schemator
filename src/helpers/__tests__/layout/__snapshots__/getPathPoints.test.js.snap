// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPathPoints() should returns path points (foreign key in left table) 1`] = `"M254 87 C336 87 256 50 338 50"`;

exports[`getPathPoints() should returns path points (foreign key in right table) 1`] = `"M338 87 C256 87 336 50 272 50"`;
